server-spi/src/main/java/org/keycloak/models/KeycloakContext.java

```diff
@@ -104,4 +104,8 @@ default Locale resolveLocale(UserModel user, Theme.Type themeType) {
     void setHttpRequest(HttpRequest httpRequest);
 
     void setHttpResponse(HttpResponse httpResponse);
+
+    UserSessionModel getUserSession();
```

Comment:
We already have the `realm`, `client`, and `authenticationSession` available from the context. It should be useful to also expose the `UserSessionModel` associated with the current session/request.

In this particular PR, we use the user session to resolve client scopes from authenticated client sessions.